spec/components/presence_channel_spec.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'presence_channel'
+
+describe PresenceChannel do
+  it "can perform basic channel functionality" do
+    # 10ms timeout for testing
+    channel1 = PresenceChannel.new("test")
+    channel2 = PresenceChannel.new("test")
+    channel3 = PresenceChannel.new("test")
+
+    channel1.clear
+
+    expect(channel3.user_ids).to eq([])
+
+    channel1.enter(user_id: 1, client_id: 1)
+    channel2.enter(user_id: 1, client_id: 2)
+
+    expect(channel3.user_ids).to eq([1])
+    expect(channel3.count).to eq(1)
+
+    channel1.leave(user_id: 1, client_id: 2)
+
+    expect(channel3.user_ids).to eq([1])
+    expect(channel3.count).to eq(1)
+
+    channel2.leave(user_id: 1, client_id: 1)
+
+    expect(channel3.user_ids).to eq([])
+    expect(channel3.count).to eq(0)
+  end
+
+  it "can automatically expire users" do
+
+    channel = PresenceChannel.new("test")
+    channel.clear
+
+    channel.enter(user_id: 1)
+    channel.enter(user_id: 1, client_id: 77)
+
+    expect(channel.count).to eq(1)
+
+    freeze_time Time.zone.now + 1 + PresenceChannel::DEFAULT_TIMEOUT
+
+    expect(channel.count).to eq(0)
+  end
+
+  it "correctly sends messages to message bus" do
+    channel = PresenceChannel.new("test")
+    channel.clear
+
+    messages = MessageBus.track_publish(channel.message_bus_channel_name) do
+      channel.enter(user_id: 1, client_id: "a")
+    end
+
+    expect(messages.length).to eq(1)
+    expected = {
+      "type" => "enter",
+      "user_id" => 1,
+    }
+    expect(messages[0].data).to eq(expected)
+
+    freeze_time Time.zone.now + 1 + PresenceChannel::DEFAULT_TIMEOUT
+
+    messages = MessageBus.track_publish(channel.message_bus_channel_name) do
+      channel.auto_leave
+    end
```

Comment:
@SamSaffron this works because we're manually calling `.auto_leave` in the spec. However, in production, I'm not sure what would cause `.auto_leave` to be triggered 🤔 

An example scenario would be:

- User A starts replying to a topic
- User B is viewing the topic, and receives an `user A entered` notification on the messagebus
- User A closes their browser, without sending an explicit `leave` event
- User B will never receive a `user B left` messagebus event, even after the `timeout` has elapsed 😢 

If some other user were to start replying, or something else called `channel.count`, then things would get cleaned up. But I don't think we can rely on that.

---

One solution would be some kind of persistent ruby thread, or regular sidekiq job, which iterates through all known channels and calls `auto_leave`. For that, we would also need to maintain a list of known channels.

I'll do some thinking and see if we can come up with a cleaner solution, but just wanted to check if you had any ideas.

Reply:
Implemented in the latest commit. For the 'list of active presence channels' I've used a ranked list, where the rank is the 'next expiry time' in the channel, so it should scale pretty well. Plus, if things have already been cleaned up by enter/leave/count events, it doesn't duplicate work.